File: strategy/python/example1/src/libs/ducks/duck.py
from abc import ABC, abstractmethod, abstractproperty
from protocols.flybehavior import FlyBehavior
from protocols.quackbeahavior import QuackBehavior
from libs.behaviors.quack.quack import Quack
from libs.behaviors.fly.flywithwings import FlyWithWings

class Duck(ABC):
    flyBehavior: FlyBehavior
    quackbehavior: QuackBehavior

    @abstractmethod
    def display(self) -> None:
        ...


    def performFly(self): 
        self.flyBehavior.fly()

    def performQuack(self):
        self.quackbehavior.quack()

    def swim(self): 
        print("All of our variants can float!")

File: strategy/typescript/example1/libs/duck/duck.ts
import { FlyBehavior } from "../../interfaces/fly-behaviour";
import { QuackBehaviour } from "../../interfaces/quack-behaviour";

export abstract class Duck {
  protected abstract flyBehaviour: FlyBehavior;
  protected abstract quackBehaviour: QuackBehaviour;

  constructor() {
    console.log("Initiated duck class");
  }

  public abstract display(): void;

  public setFlyBehaviour(fb: FlyBehavior): void {
    this.flyBehaviour = fb;
  }

  public setQuackBehaviour(qb: QuackBehaviour): void {
    this.quackBehaviour = qb;
  }

  public performFly(): void {
    this.flyBehaviour.fly();
  }

  public performQuack(): void {
    this.quackBehaviour.quack();
  }

  public swim(): void {
    console.log("All ducks can swim");
  }
}
